### PR TITLE
Tobin gh91 invalid cb cases

### PR DESCRIPTION
**Pull request by tobine**
This is mostly refactoring and clean-up in preparation for greatly expanding command buffer invalidation handling. No new checks are added in this series. Next task will be iterating through new object types (such as images and buffers), making sure the dependencies are tracked and that cmd buffer invalidation  due to those dependencies is correctly handled.

### CODE REVIEW

File: layers/core_validation_types.h [hunk 1]
```diff
@@ -102,11 +108,18 @@ struct DESCRIPTOR_POOL_NODE {
 class BUFFER_NODE : public BASE_NODE {
   public:
     using BASE_NODE::in_use;
+    using BASE_NODE::cb_bindings;
```
Comment (chrisforbes) [inline layers/core_validation_types.h]: Why?

File: layers/core_validation_types.h [hunk 2]
```diff
@@ -102,11 +108,18 @@ struct DESCRIPTOR_POOL_NODE {
...
-    BUFFER_NODE() : mem(VK_NULL_HANDLE), createInfo{} { in_use.store(0); };
-    BUFFER_NODE(const VkBufferCreateInfo *pCreateInfo) : mem(VK_NULL_HANDLE), createInfo(*pCreateInfo) { in_use.store(0); };
-    BUFFER_NODE(const BUFFER_NODE &rh_obj) : mem(rh_obj.mem), createInfo(rh_obj.createInfo) { in_use.store(rh_obj.in_use.load()); };
+    BUFFER_NODE() : buffer(VK_NULL_HANDLE), mem(VK_NULL_HANDLE), createInfo{} { in_use.store(0); };
+    BUFFER_NODE(VkBuffer buff, const VkBufferCreateInfo *pCreateInfo)
+        : buffer(buff), mem(VK_NULL_HANDLE), createInfo(*pCreateInfo) {
+        in_use.store(0);
+    };
+    BUFFER_NODE(const BUFFER_NODE &rh_obj) : buffer(rh_obj.buffer), mem(rh_obj.mem), createInfo(rh_obj.createInfo) {
+        in_use.store(rh_obj.in_use.load());
```
Comment (chrisforbes) [inline layers/core_validation_types.h]: Copying (effectively) reference counts doesn't make a lot of sense. Do you mean to make this moveable instead?

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -8079,14 +8079,14 @@ CmdCopyQueryPoolResults(VkCommandBuffer commandBuffer, VkQueryPool queryPool, ui
     GLOBAL_CB_NODE *pCB = getCBNode(dev_data, commandBuffer);
 #if MTMERGESOURCE
     VkDeviceMemory mem;
-    auto cb_data = dev_data->commandBufferMap.find(commandBuffer);
+    auto cb_node = getCBNode(dev_data, commandBuffer);
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Just did this a few lines prior

File: layers/core_validation.cpp [hunk 4]
```diff
@@ -5367,10 +5367,10 @@ FreeCommandBuffers(VkDevice device, VkCommandPool commandPool, uint32_t commandB
     std::unique_lock<std::mutex> lock(global_lock);
 
     for (uint32_t i = 0; i < commandBufferCount; i++) {
-        auto cb_pair = dev_data->commandBufferMap.find(pCommandBuffers[i]);
+        auto cb_node = getCBNode(dev_data, pCommandBuffers[i]);
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Only reason I hadn't done this already is that getCBNode is a bit weird -- it produces validation errors.

File: layers/core_validation.cpp [hunk 5]
```diff
@@ -5522,17 +5484,22 @@ VKAPI_ATTR VkResult VKAPI_CALL ResetFences(VkDevice device, uint32_t fenceCount,
...
     std::unique_lock<std::mutex> lock(global_lock);
     auto fb_node = getFramebuffer(dev_data, framebuffer);
     if (fb_node) {
-        for (auto cb_node : fb_node->cb_bindings) {
-            // Set CB as invalid and record destroyed framebuffer
-            cb_node->state = CB_INVALID;
-            cb_node->destroyedFramebuffers.insert(framebuffer);
-        }
+        invalidateCommandBuffers(fb_node->cb_bindings,
+                                 {static_cast<uint64_t>(fb_node->framebuffer), VK_DEBUG_REPORT_OBJECT_TYPE_FRAMEBUFFER_EXT});
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Is this cast going to do the right thing on all targets? If so, why are we using `reinterpret_cast<uint64_t [const] &>(ndhandle)` everywhere else?

File: layers/core_validation_types.h (hunk 1 above)
Comment (tobine) [inline layers/core_validation_types.h]: At least 10 of the objects listed in #91 will need tracking for command buffer invalidation so adding to BASE_NODE and will make state structs for each of those objects inherit from BASE_NODE to standardize handling of CB invalidation.

File: layers/core_validation.cpp (hunk 5 above)
Comment (tobine) [inline layers/core_validation.cpp]: No, needs to be reinterpret_cast. Not sure why I even put in static_cast, but I happened to only be testing on arm7 when I did and it happens to work there.

File: layers/core_validation_types.h (hunk 2 above)
Comment (tobine) [inline layers/core_validation_types.h]: I agree. I was thinking moveable when I copied the reference counts.

**Comment by tobine** [comment]
I pushed some fixes and ran this on our test farm where all configs are passing.

File: layers/core_validation_types.h (hunk 1 above)
Comment (chrisforbes) [inline layers/core_validation_types.h]: My point is there's no templates in play here to complicate name resolution -- you can refer to members from the base just fine without `using`.

**Comment by tobine** [comment]
Killed the unneeded "using" line and pushed to master.

**Pull request closed without merging**